### DISCUSSION TITLE
How to access user storage in another thread or with run.io_bound()

### DISCUSSION THREAD

**Discussion by TotoWang-hhh** [zauberzeug/nicegui#2228] [Q&A]
Not able to access the user storage in another thread. The console says that I must specify the `storage_secret`, but it has been done in the main thread.

Example:
```python
from nicegui import ui,app
import threading

def func():
    app.storage.user['x']='y'

func_t=threading.Thread(target=func)

ui.button('Run func',on_click=func_t.start)

ui.run(storage_secret='secret')

```

**Comment by rodja** [comment]
Thanks for bringing this up. For me your code produces the error

```
RuntimeError: The current slot cannot be determined because the slot stack for this task is empty.
This may happen if you try to create UI from a background task.
To fix this, enter the target slot explicitly using `with container_element:`
```

Which is already unfortunate. When fixing this like this:

```py
from nicegui import ui,app
import threading

def func():
    with content:
        app.storage.user['x']='y'

func_t=threading.Thread(target=func)

with ui.column() as content:
    ui.button('Run func',on_click=func_t.start)

ui.run(storage_secret='secret')
```

I get the error you have mentioned:

```
RuntimeError: app.storage.user needs a storage_secret passed in ui.run()
```

**Comment by rodja** [comment]
I was eager to fix this so I've created https://github.com/zauberzeug/nicegui/pull/2229. But then I came to the conclusion that we should not allow manipulation of the UI or storage from different threads. This is against the philosophy of async/await with it's main loop where exclusive access is guaranteed. Also `app.storage` uses `PersistentDict` which is not thread-safe.

**Reply by TotoWang-hhh** [reply]
So, how to solve the problem of the first error you mentioned?
```
RuntimeError: The current slot cannot be determined because the slot stack for this task is empty.
This may happen if you try to create UI from a background task.
To fix this, enter the target slot explicitly using `with container_element:`
```

**Reply by rodja** [reply]
You don't. The error message is not great and I hope we can improve it in later versions. But you should rewrite your code to do the actual storage writing outside of the background thread. Something like this:

```py
from nicegui import ui, app, run
import threading

def read_data():
    time.sleep(1) # some io bound stuff like file or web access
    return 'y'

async def update_storage():
    app.storage.user['x'] = await run.io_bound(read_data)

ui.button('Run func',on_click=update_storage)

ui.run(storage_secret='secret')
```

**Reply by TotoWang-hhh** [reply]
I have done some modifies after asking here, and it is now showing me this error

![image](https://github.com/zauberzeug/nicegui/assets/67090055/3cc4a5fd-1275-4ec9-9750-88469608bac1)

**Reply by rodja** [reply]
Looks like your create_chat_completions is still doing UI updates.

**Comment by amks1** [comment]
Agreed that it's not a good idea to manipulate the data, but is it ok to make it work for read only?

Example code:
```python3
from nicegui import ui, app, run

def fetch_something():
    time.sleep(2)
    myvar = app.storage.client.get('myvar')
    return myvar

@ui.page('/test')
def test_page():

    async def click_me():
        response = await run.io_bound(fetch_something)
        print(response)
    
    app.storage.client['myvar'] = 'myvalue'
    ui.button('Click me', on_click=click_me)

ui.run(reload=False, show=False)
```

**Reply by rodja** [reply]
Yes, read access would be ok. But that is not easy to implement (because copying the whole storage would be inefficient).
You can simply fetch the values you need beforehand and pass them to the thread:

```py
def fetch_something(myvar):
    time.sleep(0.5)
    return myvar

@ui.page('/')
def test_page():

    async def click_me():
        response = await run.io_bound(fetch_something, app.storage.client.get('myvar'))
        print(response)

    app.storage.client['myvar'] = 'myvalue'
    ui.button('Click me', on_click=click_me)
```

**Reply by amks1** [reply]
Thanks but I'm using both Nicegui and Pywebio for interfaces and I'm trying not to modify common code too much (in my actual code there are several functions between `click_me()` and `fetch_something()` where I need to retrieve the client storage).

Here's an example of something similar in [Pywebio](https://github.com/pywebio/PyWebIO).

```python3
import pywebio.output as pwo
import pywebio.session as pws
import time, threading

def fetch_something():
    time.sleep(2)
    myvar = pws.local.myvar
    print(f'From thread: {myvar}')

def click_me():
    thread = threading.Thread(target=fetch_something)
    thread.start()

pws.local.myvar = 'myvalue'
pwo.put_button('Click me', onclick=click_me)
pws.hold()
```
 
Output: 
```
From thread: myvalue
```

It's quite common in my code to hand over some processing to background tasks that needs to refer to some data about the current client. Examples of this data are User ID, Request args from the URL, Cached User Permissions (these are stored at the beginning of the request handler).
I don't understand fully how it's implemented in Pywebio, but I think Nicegui would also benefit from this.

**Reply by evnchn** [reply]
Interoperability between NiceGUI and PyWebIO is not ensured by either libraries. That's not to say you can't try, but you'll be fighting a uphill battle. 

PyWebIO "Use synchronization instead of a callback-based method to get input". Well I'd say with NiceGUI it's closer to callback method, with the Socket.IO doing the heavy lifting, so that's a red flag for incompatibility. 

Also, you have mentioned you don't want to modify common code, but your example code `fetch_something` has access to `pws.local.myvar` baked in, so that's tough to use in a NiceGUI application. 

I would recommend to remove data access from your functions, such that it can be called from PyWebIO and NiceGUI. If you find it difficult to manage the variables, a class may also help. GitHub copilot is also very good at refactoring so there's that.

**Reply by amks1** [reply]
> Also, you have mentioned you don't want to modify common code, but your example code `fetch_something` has access to `pws.local.myvar` baked in, so that's tough to use in a NiceGUI application.

I was hoping to use something like 
```python3
try:
    myvar = pws.local.myvar
except pywebio.exceptions.SessionNotFoundException:
    myvar = app.storage.client['myvar']
```
Without modifying anything else along the way. 
 
> I would recommend to remove data access from your functions, such that it can be called from PyWebIO and NiceGUI. If you find it difficult to manage the variables, a class may also help. GitHub copilot is also very good at refactoring so there's that.

I know I'll be able to work around this, and I'm not asking for help doing that. The only reason I'm posting it here is because I think it would be a good feature for NiceGUI to have.

**Reply by evnchn** [reply]
> I was hoping to use something like

Try your best not to do so. There's 4 lines of overhead and need `pws` and `app` to be defined for every function written that way... 😰

---

Anyways, the PyWebIO implementation is interesting. Here's how they do it:

https://github.com/pywebio/PyWebIO/blob/dev/pywebio%2Fsession%2F__init__.py#L483

https://github.com/pywebio/PyWebIO/blob/dev/pywebio%2Futils.py#L46-L53

> Dictionary accessed through properties. The instance does not maintain the underlying dictionary, but uses the callback function to get it every time it is accessed

> Data saved on object properties is saved to the underlying dictionary
> Accessing properties that do not exist in the data object returns None instead of throwing an exception.
> Cannot save the underscore start property
> Gets the dictionary representation of the object with obj._dict

I'll put this out for us to see what this could mean for NiceGUI, but I'm not very familiar with miltithreading and so I won't do anything personally...

**Reply by amks1** [reply]
> Try your best not to do so. There's 4 lines of overhead and need `pws` and `app` to be defined for every function written that way... 😰

I only need to do this in the last function, so I just need to import them, no other definition necessary.
```python3
import pywebio.session as pws
from nicegui import app
```
in just one place.

**Comment by TotoWang-hhh** [comment]
Aww come on u guys just kept calling me back to go through the piece of sh*t I wrote years ago again and again 😭

**Reply by evnchn** [reply]
🤡
